src/gpio.rs

```diff
@@ -50,1205 +46,1289 @@ pub trait GpioExt {
+        fn medium(&mut self, i: u8);
+        fn high(&mut self, i: u8);
+    }
+
+    pub trait Pupdr {
+        fn floating(&mut self, i: u8);
+        fn pull_up(&mut self, i: u8);
+        fn pull_down(&mut self, i: u8);
+    }
+
+    pub trait Afr {
+        fn afx(&mut self, i: u8, x: u8);
+    }
+
+    pub trait Gpio {
+        type Reg: GpioRegExt + ?Sized;
+
+        fn ptr(&self) -> *const Self::Reg;
+    }
 }
 
-/// Floating input (type state)
-pub struct Floating;
-/// Pulled down input (type state)
-pub struct PullDown;
-/// Pulled up input (type state)
-pub struct PullUp;
-
-/// Output mode (type state)
-pub struct Output<MODE> {
-    _mode: PhantomData<MODE>,
+use private::{Afr, GpioRegExt, Moder, Ospeedr, Otyper, Pupdr};
+
+/// Marker trait for GPIO ports
+pub trait Gpio: private::Gpio {}
+
+/// Marker trait for compile time defined GPIO ports
+pub trait GpioStatic: Gpio {
+    /// Associated MODER register
+    type MODER: Moder;
+    /// Associated OTYPER register
+    type OTYPER: Otyper;
+    /// Associated OSPEEDR register
+    type OSPEEDR: Ospeedr;
+    /// Associated PUPDR register
+    type PUPDR: Pupdr;
 }
 
-/// Push pull output (type state)
-pub struct PushPull;
-/// Open drain output (type state)
-pub struct OpenDrain;
+/// Marker trait for pin number
+pub trait Index {
+    #[doc(hidden)]
+    fn index(&self) -> u8;
+}
 
-/// Analog mode (type state)
-pub struct Analog;
+impl<U> Index for U
+where
+    U: Unsigned,
+{
+    #[inline(always)]
+    fn index(&self) -> u8 {
+        Self::U8
+    }
+}
 
-/// Alternate function 0 (type state)
-pub struct AF0;
+/// Marker trait for readable pin modes
+pub trait Readable {}
 
-/// Alternate function 1 (type state)
-pub struct AF1;
+/// Marker trait for slew rate configurable pin modes
+pub trait OutputSpeed {}
 
-/// Alternate function 2 (type state)
-pub struct AF2;
+/// Marker trait for internal resistor enabled pin modes
+pub trait InternalResistor {}
 
-/// Alternate function 3 (type state)
-pub struct AF3;
+/// Runtime defined GPIO port (type state)
+pub struct Gpiox {
+    ptr: *const dyn GpioRegExt,
+}
 
-/// Alternate function 4 (type state)
-pub struct AF4;
+impl private::Gpio for Gpiox {
+    type Reg = dyn GpioRegExt;
 
-/// Alternate function 5 (type state)
-pub struct AF5;
+    fn ptr(&self) -> *const Self::Reg {
+        self.ptr
+    }
+}
 
-/// Alternate function 6 (type state)
-pub struct AF6;
+impl Gpio for Gpiox {}
 
-/// Alternate function 7 (type state)
-pub struct AF7;
+/// Runtime defined pin number (type state)
+pub struct Ux(u8);
 
-/// Alternate function 8 (type state)
-pub struct AF8;
+impl Index for Ux {
+    fn index(&self) -> u8 {
+        self.0
+    }
+}
 
-/// Alternate function 9 (type state)
-pub struct AF9;
+/// Input mode (type state)
+pub struct Input;
+/// Output mode (type state)
+pub struct Output<OTYPE>(PhantomData<OTYPE>);
+/// Alternate function (type state)
+pub struct Alternate<AF, OTYPE>(PhantomData<AF>, PhantomData<OTYPE>);
```

Comment:
I've noticed many changes, where `Alternate<PushPull>` is now enforced. 
Is this really needed for every case? [`stm32f4xx-hal`](https://github.com/stm32-rs/stm32f4xx-hal/blob/dadf2192e6ade2652de3b02754c3606532dd624d/src/gpio.rs#L34-L42) does only differentiate between `Alternate` and `Alternate<OpenDrain>` / `AlternateOD`. 

Is there a use case for a neutral `Alternate` Type State, where it does not matter, which Output State is used?
Point 3. in https://github.com/stm32-rs/stm32f3xx-hal/issues/125#issue-665855804 might be a problem, right? 

What is your thought about this?

Reply:
I follow the [stm32f1xx-hal](https://github.com/stm32-rs/stm32f1xx-hal/blob/ca01418e356d2879628dd86cc4c63597d85c1665/src/gpio.rs#L484-L530) as it feels more consistent.

Some peripherals should accept `Alternate<MODE>` instead of `Alternate<PushPull>`, maybe?

Reply:
Support for OD pins is added for serial and pwm peripherals.